src/stream/src/executor/subscription.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::ops::Bound;
+
+use either::Either;
+use futures::pin_mut;
+use futures::prelude::stream::{select_with_strategy, PollNext, StreamExt};
+use futures_async_stream::try_stream;
+use risingwave_common::catalog::Schema;
+use risingwave_common::field_generator::ChronoFieldInner;
+use risingwave_common::row::{self, OwnedRow, Row};
+use risingwave_common::types::{ScalarImpl, Timestamptz};
+use risingwave_connector::sink::log_store::TruncateOffset;
+use risingwave_storage::store::LocalStateStore;
+use risingwave_storage::StateStore;
+use tokio::time::sleep;
+
+use super::test_utils::prelude::StateTable;
+use super::{
+    expect_first_barrier, ActorContextRef, BoxedExecutor, BoxedMessageStream, Executor,
+    ExecutorInfo, Message, PkIndicesRef, StreamExecutorError, StreamExecutorResult,
+};
+use crate::common::log_store_impl::only_writer_log_store::OnlyWriterLogStoreWriter;
+
+const EXECUTE_GC_INTERVAL: u64 = 3600;
+
+pub struct SubscriptionExecutor<LS: LocalStateStore, S: StateStore> {
+    actor_context: ActorContextRef,
+    info: ExecutorInfo,
+    input: BoxedExecutor,
+    log_store: OnlyWriterLogStoreWriter<LS>,
+    epoch_store: StateTable<S>,
```

Comment:
I think we can directly use epoch as the timestamp instead of maintaining a map from current real timestamp to epoch.

Reply:
Considered this approach, but since the epoch is generated from the meta, there may be a significant delay time to reach this operator, and in extreme cases this may result in a particular epoch surviving in the log store for too short a period of time, so would like to compute the survival time when determining the arrival of this epoch at the operator

Reply:
> there may be a significant delay time

Usually the ttl should be set as several days or even months, but by design in our system, delay of hour or even minute is not expected, and even if this rare unexpected thing happens, compared to the days or months ttl, this delay is small and can be ignored.

More importantly, we should have multiple parallelisms instead of singleton, but in current implementation, the `epoch_store` writes `(timestamp, epoch)` to the store, and there will be multiple different value of timestamp for a single epoch, and this can be ambiguous. Even if we write one record for each vnode, it will still be strange for an epoch to have different timestamp value in different vnodes. If we want to unify the timestamp in all vnodes, the coordination between different parallelisms will be complicated. 

Given that it will be more complicated to support this feature but it has little benefit, I think we don't have to store the timestamp of each epoch. We can directly use the epoch value as the timestamp.